src/os_auth/auth.c

```diff
@@ -0,0 +1,326 @@
+        node->group = strdup(group);
+
+    (*insert_tail) = node;
+    insert_tail = &node->next;
+}
+
+// Append key to backup queue
+void add_backup(const keyentry *entry) {
+    struct keynode *node;
+
+    os_calloc(1, sizeof(struct keynode), node);
+    node->id = strdup(entry->id);
+    node->name = strdup(entry->name);
+    node->ip = strdup(entry->ip->ip);
+
+    (*backup_tail) = node;
+    backup_tail = &node->next;
+}
+
+// Append key to deletion queue
+void add_remove(const keyentry *entry) {
+    struct keynode *node;
+
+    os_calloc(1, sizeof(struct keynode), node);
+    node->id = strdup(entry->id);
+    node->name = strdup(entry->name);
+    node->ip = strdup(entry->ip->ip);
+
+    (*remove_tail) = node;
+    remove_tail = &node->next;
+}
+
+
+w_err_t w_auth_parse_data(const char* buf, char *response,const char *authpass, char *ip, char **agentname, char **groups){
+    
+    bool parseok = FALSE;     
+    /* Checking for shared password authentication. */
+    if(authpass) {
+        /* Format is pretty simple: OSSEC PASS: PASS WHATEVERACTION */
+        parseok = FALSE; 
+        if (strncmp(buf, "OSSEC PASS: ", 12) == 0) {
+            buf += 12;
+            if (strlen(buf) > strlen(authpass) && strncmp(buf, authpass, strlen(authpass)) == 0) {
+                buf += strlen(authpass);
+                if (*buf == ' ') {
+                    buf++;
+                    parseok = 1;
+                }
+            }
+        }
+
+        if (parseok == 0) {
+            merror("Invalid password provided by %s. Closing connection.", ip);
+            snprintf(response, 2048, "ERROR: Invalid password\n\n");  
+            return OS_INVALID;
+        }
+    }
+
+    /* Checking for action A (add agent) */    
+    parseok = FALSE;
+    if (strncmp(buf, "OSSEC A:'", 9) == 0) {        
+        buf += 9;
+                
+        unsigned len = 0;
+        while (*buf != '\0') {
+            if (*buf == '\'') {
+                os_malloc(len+1, *agentname);
+                memcpy(*agentname, buf-len, len);
+                (*agentname)[len] = '\0';
+                minfo("Received request for a new agent (%s) from: %s", *agentname, ip);
+                parseok = TRUE;
+                break;
+            }
+            len++; 
+            buf++;           
+        }
+    }
+    buf++;    
+    
+    if (!parseok) {
+        merror("Invalid request for new agent from: %s", ip);
+        snprintf(response, 2048, "ERROR: Invalid request for new agent\n\n");  
+        return OS_INVALID;
+    }    
+
+    if (!OS_IsValidName(*agentname)) {
+        merror("Invalid agent name: %s from %s", *agentname, ip);
+        snprintf(response, 2048, "ERROR: Invalid agent name: %s\n\n", *agentname);  
+        return OS_INVALID;
+    }
+
+    /* Check for valid centralized group */
+    
+    char centralized_group_token[2] = "G:";
+
+    if(strncmp(++buf,centralized_group_token,2)==0)
+    {
+        char tmp_groups[OS_SIZE_65536] = {0};
```

Comment:
Either group length should be checked before this truncation or we should use a bigger buffer (OS_SIZE_65536 + OS_SIZE_4096 + 1) to properly identify when multi-group string is higher than allowed.

Reply:
As sscanf was truncating the string, tmp_groups will be now OS_SIZE_65536+1, and sscanf will read 65536 bytes wich is the complete size assigned for groups.